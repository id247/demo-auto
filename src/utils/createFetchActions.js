export default (prefix, name) => {
  const requestType = `${prefix}${name}_REQUEST`;
  const successType = `${prefix}${name}_SUCCESS`;
  const failType = `${prefix}${name}_FAIL`;

  return {
    requestType,
    successType,
    failType,
    fetch(payload) {
      return {
        type: requestType,
        payload
      };
    },
    success(payload) {
      return {
        type: successType,
        payload
      };
    },
    fail(payload) {
      return {
        type: failType,
        payload
      };
    }
  };
};
