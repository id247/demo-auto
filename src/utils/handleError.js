import history from 'myHistory';

export const handleError = e => {
  if (!e) return;

  if (e.response && e.response.status === 404) {
    history.replace('/404');
  } else {
    console.error(e);
  }
};

export default handleError;
