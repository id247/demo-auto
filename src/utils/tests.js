import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import createStore from 'store';

export const testActionCreator = ({ type, fn }) => {
  describe(fn.name, () => {
    it('should return correct action', () => {
      expect(fn({})).toEqual({ type, payload: {} });
    });

    it('should return undefined payload if non was passed', () => {
      expect(fn()).toEqual({ type, payload: undefined });
    });
  });
};

export const renderWithReduxAndRouter = (
  ui,
  {
    initialState,
    store = createStore(initialState),
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <Router history={history}>{ui}</Router>
    </Provider>
  ),
  store,
  history
});
