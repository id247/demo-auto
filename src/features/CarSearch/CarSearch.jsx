import React from 'react';
import qs from 'query-string';
import PropTypes from 'prop-types';
import { fetchStatesPropType } from 'constants/propTypes';
import FetchError from 'components/FetchError';
import CarSearchLayout from 'components/CarSearchLayout';
import CarSearchInfo from 'features/CarSearchInfo';
import Pagination from 'components/Pagination';
import CarList from 'components/CarList';
import CarListItem from 'features/CarListItem';
import Filters from 'features/Filters';
import { omit } from 'utils/object';

class CarSearch extends React.Component {
  componentDidMount() {
    this.fetchData();
  }

  componentDidUpdate(prevProps) {
    const { queryString, fetchStates } = this.props;
    if (prevProps.queryString !== queryString) {
      this.fetchData();
    }

    if (fetchStates.isFetchingDone && prevProps.fetchStates.isFetching) {
      // TODO add smoth scloll
      window && window.scroll(0, 0);
    }
  }

  fetchData = () => {
    const { queryParams, fetchCarsAsync } = this.props;

    fetchCarsAsync({ queryParams });
  };

  updateUrl = newParams => {
    const { history, queryParams } = this.props;
    const newQueryParams = {
      ...omit(queryParams, 'page'),
      ...newParams
    };
    history.push(`?${qs.stringify(newQueryParams)}`);
  };

  render() {
    const { carIds, fetchStates, totalItems, queryParams = {} } = this.props;

    return (
      <CarSearchLayout
        sidebar={
          <Filters updateUrl={this.updateUrl} queryParams={queryParams} />
        }
        info={
          <CarSearchInfo
            fetchStates={fetchStates}
            updateUrl={this.updateUrl}
            queryParams={queryParams}
            itemsToDisplay={carIds.length}
            totalItems={totalItems}
          />
        }
      >
        {fetchStates.isFetchingError ? (
          <FetchError fetchData={this.fetchData} />
        ) : (
          <CarList
            carIds={carIds}
            fetchStates={fetchStates}
            renderCarListItem={({ carId }) => (
              <CarListItem carId={carId} isIdle={fetchStates.isIdle} />
            )}
            pagination={
              <Pagination
                queryParams={queryParams}
                currentPage={parseInt(queryParams.page, 10) || 1}
                totalItems={totalItems}
              />
            }
          />
        )}
      </CarSearchLayout>
    );
  }
}

CarSearch.propTypes = {
  queryString: PropTypes.string.isRequired,
  queryParams: PropTypes.object.isRequired,
  fetchStates: fetchStatesPropType,
  history: PropTypes.object.isRequired,
  carIds: PropTypes.arrayOf(PropTypes.number).isRequired,
  fetchCarsAsync: PropTypes.func.isRequired,
  totalItems: PropTypes.number.isRequired
};

export default CarSearch;
