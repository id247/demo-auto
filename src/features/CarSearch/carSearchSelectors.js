import { createSelector } from 'reselect';
import { carsSelector } from 'selectors/cars';
import { getFetchStates } from 'utils/stateMachine';

export const carSearchSelector = state => state.carSearch;

export const carSearchIdsSelector = createSelector(
  carSearchSelector,
  carSearch => carSearch.ids
);

export const carSearchCarsSelector = createSelector(
  carsSelector,
  carSearchIdsSelector,
  (cars, carIds) => carIds.map(c => cars[c])
);

export const carSearchTotalItemsSelector = createSelector(
  carSearchSelector,
  carSearch => carSearch.totalItems
);

export const carSearchFetchStatesSelector = createSelector(
  carSearchSelector,
  carSearch => getFetchStates(carSearch.fetchState)
);
