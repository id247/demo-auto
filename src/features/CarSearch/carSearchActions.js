import apiRequest from 'api';
import apiRoutes from 'api/api-routes';
import { adaptCars, adaptCarIds } from 'adapters/cars';
import { addCars } from 'actions/cars';
import createFetchActions from 'utils/createFetchActions';
import handleError from 'utils/handleError';
import { defaultPerPage } from 'constants/pagination';

const prefix = 'carSearch/';

export const fetchCarsActions = createFetchActions(prefix, 'FETCH_CARS');

const paramsMap = {
  page: { key: '_page', default: 1 },
  limit: { key: '_limit', default: defaultPerPage }
};

const defaultParams = Object.values(paramsMap).reduce((acc, v) => {
  acc[v.key] = v.default;
  return acc;
}, {});

export const fetchCarsAsync = ({ queryParams = {} }) => async dispatch => {
  dispatch(fetchCarsActions.fetch());
  try {
    const fullQueryParams = Object.keys(queryParams).reduce((acc, k) => {
      if (paramsMap[k]) {
        acc[paramsMap[k].key] = queryParams[k];
      } else {
        acc[k] = queryParams[k];
      }

      return acc;
    }, defaultParams);

    const res = await apiRequest(apiRoutes.cars, {
      params: fullQueryParams
    });

    const { totalItems, items: cars } = res;

    if (totalItems > 0 && cars.length === 0) {
      const error = new Error('No cars found');
      error.response = {
        status: 404
      };
      throw error;
    }

    dispatch(addCars({ cars: adaptCars(cars) }));
    dispatch(
      fetchCarsActions.success({ carIds: adaptCarIds(cars), totalItems })
    );
  } catch (e) {
    dispatch(fetchCarsActions.fail());
    handleError(e);
  }
};
