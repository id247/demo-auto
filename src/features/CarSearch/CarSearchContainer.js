import { connect } from 'react-redux';
import {
  carSearchIdsSelector,
  carSearchTotalItemsSelector,
  carSearchFetchStatesSelector
} from './carSearchSelectors';
import { fetchCarsAsync } from './carSearchActions';
import CarSearch from './CarSearch';

export default connect(
  state => ({
    carIds: carSearchIdsSelector(state),
    totalItems: carSearchTotalItemsSelector(state),
    fetchStates: carSearchFetchStatesSelector(state)
  }),
  {
    fetchCarsAsync
  }
)(CarSearch);
