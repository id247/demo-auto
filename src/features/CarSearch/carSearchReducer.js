import fetchStates from 'constants/fetchStates';
import { getNewStateMachineState } from 'utils/stateMachine';
import { fetchCarsActions } from './carSearchActions';

const stateMachine = {
  [fetchStates.idle]: {
    name: fetchStates.idle,
    to: {
      [fetchCarsActions.successType]: fetchStates.done,
      [fetchCarsActions.failType]: fetchStates.error
    }
  },
  [fetchStates.fetching]: {
    name: fetchStates.fetching,
    to: {
      [fetchCarsActions.successType]: fetchStates.done,
      [fetchCarsActions.failType]: fetchStates.error
    }
  },
  [fetchStates.done]: {
    name: fetchStates.done,
    to: {
      [fetchCarsActions.requestType]: fetchStates.fetching
    }
  },
  [fetchStates.error]: {
    name: fetchStates.error,
    to: {
      [fetchCarsActions.requestType]: fetchStates.idle
    }
  }
};

export const initialState = {
  ids: [],
  totalItems: 0,
  fetchState: fetchStates.idle
};

const carSearchReducer = (state = initialState, { type, payload }) => {
  const fetchState = getNewStateMachineState(
    stateMachine,
    state.fetchState,
    type
  );

  switch (type) {
    case fetchCarsActions.requestType:
    case fetchCarsActions.failType:
      return { ...state, fetchState };
    case fetchCarsActions.successType:
      return {
        ...state,
        ids: payload.carIds,
        totalItems: payload.totalItems,
        fetchState
      };
    default:
      return state;
  }
};

export default carSearchReducer;
