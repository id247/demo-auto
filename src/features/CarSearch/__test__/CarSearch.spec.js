import 'jest-dom/extend-expect';
import React from 'react';
import { cleanup, wait } from 'react-testing-library';
import axios from 'axios';
import { renderWithReduxAndRouter } from 'utils/tests';
import CarSearch from '../';

afterEach(() => {
  //eslint-disable-next-line no-console
  console.error = jest.fn();
  cleanup();
  axios.mockReset();
});

describe('CarSearch component', () => {
  const carBE = {
    color: 'red',
    fuelType: 'Diesel',
    manufacturerName: 'Mercedes-Benz',
    mileage: { number: 158099, unit: 'km' },
    unit: 'km',
    modelName: 'A-Klasse',
    pictureUrl: 'http://localhost/car.jpg',
    id: 10252,
    text: 'description'
  };
  const cars = [carBE];
  const totalItems = 1;

  const props = {
    queryString: '',
    queryParams: {},
    fetchStates: {
      isIdle: true,
      isFetching: false,
      isFetchingDone: false,
      isFetchingError: false
    },
    history: {
      push: jest.fn()
    }
  };

  test('Initial idle render', async () => {
    axios.mockImplementation(o => {
      let res = {};
      if (o.url.endsWith('/colors')) {
        res = { data: ['white'] };
      } else if (o.url.endsWith('/manufacturers')) {
        res = { data: [{ name: 'Mercedes' }] };
      } else {
        res = { data: cars, headers: { 'x-total-count': totalItems } };
      }
      return Promise.resolve(res);
    });

    const { container } = renderWithReduxAndRouter(<CarSearch {...props} />);

    expect(container).toMatchSnapshot();
  });

  test('Render after successful fetching ', async () => {
    axios.mockImplementation(o => {
      let res = {};
      if (o.url.endsWith('/colors')) {
        res = { data: ['white'] };
      } else if (o.url.endsWith('/manufacturers')) {
        res = { data: [{ name: 'Mercedes' }] };
      } else {
        res = { data: cars, headers: { 'x-total-count': totalItems } };
      }
      return Promise.resolve(res);
    });

    const { container, getByText } = renderWithReduxAndRouter(
      <CarSearch {...props} />
    );

    expect(axios).toHaveBeenCalledTimes(3);

    await wait(() => getByText('Mercedes-Benz A-Klasse'));

    expect(container).toMatchSnapshot();
  });

  test('Render after unsuccessful fetching', async () => {
    axios.mockImplementation(o => {
      let data = {};
      if (o.url.endsWith('/colors')) {
        data = { data: ['white'] };
      } else if (o.url.endsWith('/manufacturers')) {
        data = { data: [{ name: 'Mercedes' }] };
      } else {
        return Promise.reject();
      }
      return Promise.resolve({ data });
    });

    const { container, getByText } = renderWithReduxAndRouter(
      <CarSearch {...props} />
    );

    expect(axios).toHaveBeenCalledTimes(3);

    await wait(() => getByText('Something went wrong'));

    expect(container).toMatchSnapshot();
  });
});
