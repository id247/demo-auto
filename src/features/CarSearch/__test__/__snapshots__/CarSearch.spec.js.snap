// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarSearch component Initial idle render 1`] = `
<div>
  <div
    class="wrapper wrapper--l"
  >
    <div
      class="car-search-layout"
    >
      <div
        class="car-search-layout__sidebar"
      >
        <div
          class="sidebar-item"
        >
          <div
            class="form"
          >
            <div
              class="form__line"
            >
              <div
                class="form__label"
              >
                <span
                  class="idle"
                >
                   
                </span>
              </div>
              <div
                class="form__input"
              >
                <span
                  class="idle"
                  style="height: 20px;"
                >
                   
                </span>
              </div>
            </div>
            <div
              class="form__line"
            >
              <div
                class="form__label"
              >
                <span
                  class="idle"
                >
                   
                </span>
              </div>
              <div
                class="form__input"
              >
                <span
                  class="idle"
                  style="height: 20px;"
                >
                   
                </span>
              </div>
            </div>
            <div
              class="form__buttons"
            >
              <div
                class="sidebar-button"
              >
                <span
                  class="idle"
                  style="display: inline-block; width: 128px; height: 30px;"
                >
                   
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="car-search-layout__content"
      >
        <div
          class="car-search-layout__info"
        >
          <div
            class="car-search-info"
          >
            <div
              class="car-search-info__content"
            >
              <div
                class="car-search-info__title"
              >
                Available cars
              </div>
              <div
                class="car-search-info__results"
              >
                <span
                  class="idle"
                  style="width: 100px;"
                >
                   
                </span>
              </div>
            </div>
            <div
              class="car-search-info__sort"
            >
              <form
                class="form"
              >
                <div
                  class="form__line"
                >
                  <label
                    class="form__label"
                    for="sort"
                  >
                    Sort By
                  </label>
                  <div
                    class="form__input"
                  >
                    <div
                      class="select"
                      data-testid="select"
                      id="sort"
                      name="sort"
                    >
                      <button
                        class="select__button"
                        data-testid="select-button"
                        type="button"
                      >
                        None
                      </button>
                      <input
                        name="sort"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div
          class="car-search-layout__list"
        >
          <div
            class="car-list"
          >
            <div
              class="car-list__list"
            >
              <div
                class="car-list__item"
              >
                <div
                  class="car-list-item"
                >
                  <div
                    class="car-list-item__image-wrapper"
                  >
                    <span
                      class="idle"
                      style="width: 100%;"
                    >
                       
                    </span>
                  </div>
                  <div
                    class="car-list-item__content"
                  >
                    <h3
                      class="car-list-item__title"
                    >
                      <span
                        class="idle"
                        style="width: 50%;"
                      >
                         
                      </span>
                    </h3>
                    <div
                      class="car-list-item__info"
                    >
                      <span
                        class="idle"
                        style="width: 70%; margin-bottom: 5px;"
                      >
                         
                      </span>
                      <span
                        class="idle"
                        style="width: 70%;"
                      >
                         
                      </span>
                    </div>
                    <span
                      class="idle"
                      style="width: 60px;"
                    >
                       
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="car-list__item"
              >
                <div
                  class="car-list-item"
                >
                  <div
                    class="car-list-item__image-wrapper"
                  >
                    <span
                      class="idle"
                      style="width: 100%;"
                    >
                       
                    </span>
                  </div>
                  <div
                    class="car-list-item__content"
                  >
                    <h3
                      class="car-list-item__title"
                    >
                      <span
                        class="idle"
                        style="width: 50%;"
                      >
                         
                      </span>
                    </h3>
                    <div
                      class="car-list-item__info"
                    >
                      <span
                        class="idle"
                        style="width: 70%; margin-bottom: 5px;"
                      >
                         
                      </span>
                      <span
                        class="idle"
                        style="width: 70%;"
                      >
                         
                      </span>
                    </div>
                    <span
                      class="idle"
                      style="width: 60px;"
                    >
                       
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="car-list__item"
              >
                <div
                  class="car-list-item"
                >
                  <div
                    class="car-list-item__image-wrapper"
                  >
                    <span
                      class="idle"
                      style="width: 100%;"
                    >
                       
                    </span>
                  </div>
                  <div
                    class="car-list-item__content"
                  >
                    <h3
                      class="car-list-item__title"
                    >
                      <span
                        class="idle"
                        style="width: 50%;"
                      >
                         
                      </span>
                    </h3>
                    <div
                      class="car-list-item__info"
                    >
                      <span
                        class="idle"
                        style="width: 70%; margin-bottom: 5px;"
                      >
                         
                      </span>
                      <span
                        class="idle"
                        style="width: 70%;"
                      >
                         
                      </span>
                    </div>
                    <span
                      class="idle"
                      style="width: 60px;"
                    >
                       
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="car-list__item"
              >
                <div
                  class="car-list-item"
                >
                  <div
                    class="car-list-item__image-wrapper"
                  >
                    <span
                      class="idle"
                      style="width: 100%;"
                    >
                       
                    </span>
                  </div>
                  <div
                    class="car-list-item__content"
                  >
                    <h3
                      class="car-list-item__title"
                    >
                      <span
                        class="idle"
                        style="width: 50%;"
                      >
                         
                      </span>
                    </h3>
                    <div
                      class="car-list-item__info"
                    >
                      <span
                        class="idle"
                        style="width: 70%; margin-bottom: 5px;"
                      >
                         
                      </span>
                      <span
                        class="idle"
                        style="width: 70%;"
                      >
                         
                      </span>
                    </div>
                    <span
                      class="idle"
                      style="width: 60px;"
                    >
                       
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="car-list__item"
              >
                <div
                  class="car-list-item"
                >
                  <div
                    class="car-list-item__image-wrapper"
                  >
                    <span
                      class="idle"
                      style="width: 100%;"
                    >
                       
                    </span>
                  </div>
                  <div
                    class="car-list-item__content"
                  >
                    <h3
                      class="car-list-item__title"
                    >
                      <span
                        class="idle"
                        style="width: 50%;"
                      >
                         
                      </span>
                    </h3>
                    <div
                      class="car-list-item__info"
                    >
                      <span
                        class="idle"
                        style="width: 70%; margin-bottom: 5px;"
                      >
                         
                      </span>
                      <span
                        class="idle"
                        style="width: 70%;"
                      >
                         
                      </span>
                    </div>
                    <span
                      class="idle"
                      style="width: 60px;"
                    >
                       
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CarSearch component Render after successful fetching  1`] = `
<div>
  <div
    class="wrapper wrapper--l"
  >
    <div
      class="car-search-layout"
    >
      <div
        class="car-search-layout__sidebar"
      >
        <div
          class="sidebar-item"
        >
          <form
            class="form"
          >
            <div
              class="form__line"
            >
              <label
                class="form__label"
                for="color"
              >
                Color
              </label>
              <div
                class="form__input"
              >
                <div
                  class="select"
                  data-testid="select"
                  id="color"
                  name="color"
                >
                  <button
                    class="select__button"
                    data-testid="select-button"
                    type="button"
                  >
                    All car colors
                  </button>
                  <input
                    name="color"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="form__line"
            >
              <label
                class="form__label"
                for="manufacturer"
              >
                Manufacturer
              </label>
              <div
                class="form__input"
              >
                <div
                  class="select"
                  data-testid="select"
                  id="manufacturer"
                  name="manufacturer"
                >
                  <button
                    class="select__button"
                    data-testid="select-button"
                    type="button"
                  >
                    All manufacturers
                  </button>
                  <input
                    name="manufacturer"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="form__buttons"
            >
              <div
                class="sidebar-button"
              >
                <button
                  class="button button--primary"
                  data-testid="button"
                  type="submit"
                >
                  Filter
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
      <div
        class="car-search-layout__content"
      >
        <div
          class="car-search-layout__info"
        >
          <div
            class="car-search-info"
          >
            <div
              class="car-search-info__content"
            >
              <div
                class="car-search-info__title"
              >
                Available cars
              </div>
              <div
                class="car-search-info__results"
              >
                Showing 1 of 1 results
              </div>
            </div>
            <div
              class="car-search-info__sort"
            >
              <form
                class="form"
              >
                <div
                  class="form__line"
                >
                  <label
                    class="form__label"
                    for="sort"
                  >
                    Sort By
                  </label>
                  <div
                    class="form__input"
                  >
                    <div
                      class="select"
                      data-testid="select"
                      id="sort"
                      name="sort"
                    >
                      <button
                        class="select__button"
                        data-testid="select-button"
                        type="button"
                      >
                        None
                      </button>
                      <input
                        name="sort"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div
          class="car-search-layout__list"
        >
          <div
            class="car-list"
          >
            <div
              class="car-list__list"
            >
              <div
                class="car-list__item"
              >
                <div
                  class="car-list-item"
                >
                  <div
                    class="car-list-item__image-wrapper"
                  >
                    <img
                      alt="Mercedes-Benz A-Klasse"
                      class="car-list-item__image"
                    />
                  </div>
                  <div
                    class="car-list-item__content"
                  >
                    <h3
                      class="car-list-item__title"
                    >
                      Mercedes-Benz A-Klasse
                    </h3>
                    <div
                      class="car-list-item__info"
                    >
                      Stock # 10252 - 158099 km - Diesel - red
                    </div>
                    <a
                      class="car-list-item__link"
                      href="/car/10252"
                    >
                      View details
                    </a>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="car-list__pagination"
            >
              <div
                class="pagination"
              >
                <div
                  class="pagination__item"
                >
                  First
                </div>
                <div
                  class="pagination__item"
                >
                  Previous
                </div>
                <div
                  class="pagination__item"
                >
                  Page 
                  1
                   of 
                  0
                </div>
                <div
                  class="pagination__item"
                >
                  <a
                    class="pagination__link"
                    href="/?page=0"
                  >
                    Next
                  </a>
                </div>
                <div
                  class="pagination__item"
                >
                  <a
                    class="pagination__link"
                    href="/?page=0"
                  >
                    Last
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CarSearch component Render after unsuccessful fetching 1`] = `
<div>
  <div
    class="wrapper wrapper--l"
  >
    <div
      class="car-search-layout"
    >
      <div
        class="car-search-layout__sidebar"
      >
        <div
          class="error"
        >
          <div
            class="error__text"
          >
            Something went wrong
          </div>
          <button
            class="button button--primary"
            data-testid="button"
            type="button"
          >
            Retry
          </button>
        </div>
      </div>
      <div
        class="car-search-layout__content"
      >
        <div
          class="car-search-layout__info"
        >
          <div
            class="car-search-info"
          >
            <div
              class="car-search-info__content"
            >
              <div
                class="car-search-info__title"
              >
                Available cars
              </div>
              <div
                class="car-search-info__results"
              >
                Showing 0 of 0 results
              </div>
            </div>
            <div
              class="car-search-info__sort"
            >
              <form
                class="form"
              >
                <div
                  class="form__line"
                >
                  <label
                    class="form__label"
                    for="sort"
                  >
                    Sort By
                  </label>
                  <div
                    class="form__input"
                  >
                    <div
                      class="select"
                      data-testid="select"
                      id="sort"
                      name="sort"
                    >
                      <button
                        class="select__button"
                        data-testid="select-button"
                        type="button"
                      >
                        None
                      </button>
                      <input
                        name="sort"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div
          class="car-search-layout__list"
        >
          <div
            class="error"
          >
            <div
              class="error__text"
            >
              Something went wrong
            </div>
            <button
              class="button button--primary"
              data-testid="button"
              type="button"
            >
              Retry
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
